bpo-44962 fix a race in WeakKeyDict, WeakValueDict and WeakSet when two threads attempt to commit the last pending removal

graingert opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-44962](https://bugs.python.org/issue44962) -->
https://bugs.python.org/issue44962
<!-- /issue-number -->

graingert committed:
bpo-44962 fix a race in WeakSet

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

graingert commented on Lib/_weakrefset.py:
```diff
@@ -51,10 +51,15 @@ def _remove(item, selfref=ref(self)):
-        l = self._pending_removals
+        pop = self._pending_removals.pop
         discard = self.data.discard
-        while l:
-            discard(l.pop())
```
the race is between the `while l:` and the `l.pop()`: There can be one pending removal in the list, `l`, and two threads passing the `while l:` guard

This PR uses EAFP to avoid that scenario

graingert commented on Misc/NEWS.d/next/Core and Builtins/2021-08-23-19-55-08.bpo-44962.J00ftt.rst:
```diff
@@ -0,0 +1 @@
+fix a race in WeakSet
```
```suggestion
Fix a race in WeakSet when two threads attempt to commit the last pending removal. This fixes asyncio.create_task and fixes a data loss in asyncio.run where shutdown_asyncgens is not run
```

graingert committed:
Update Misc/NEWS.d/next/Core and Builtins/2021-08-23-19-55-08.bpo-44962.J00ftt.rst

bensimner reviewed (commented):
The change being overriden here was introduced by @pitrou in https://github.com/python/cpython/commit/c1baa601e2b558deb690edfdf334fceee3b03327#diff-0d950b54d0d2b4cc98bbe2945b67e8e5c12ec14636861705e275aa86b35d73cf

The above change also touches WeakKeyDictionary and WeakValueDictionary too,  should those also get the updated _commit_removals change?  are there any other races that should be fixed?

graingert committed:
fix races in WeakKeyDict and WeakValueDict

graingert committed:
catch KeyError in WeakKeyDict callback

fixes:
Exception ignored in: weakref callback <function WeakKeyDictionary.__init__.<locals>.remove at 0x00007fe82245d2e0>
Traceback (most recent call last):
  File "/usr/lib/pypy3/lib-python/3/weakref.py", line 390, in remove
    del self.data[k]
KeyError: <weakref at 0x00007fe76e8d8180; dead>
Exception ignored in: weakref callback <function WeakKeyDictionary.__init__.<locals>.remove at 0x00007fe82245d2e0>
Traceback (most recent call last):
  File "/usr/lib/pypy3/lib-python/3/weakref.py", line 390, in remove
    del self.data[k]
KeyError: <weakref at 0x00007fe76e8d81a0; dead>
Exception ignored in: weakref callback <function WeakKeyDictionary.__init__.<locals>.remove at 0x00007fe82245d2e0>
Traceback (most recent call last):
  File "/usr/lib/pypy3/lib-python/3/weakref.py", line 390, in remove
    del self.data[k]
KeyError: <weakref at 0x000056548f1e24a0; dead>

see https://github.com/agronholm/anyio/issues/362#issuecomment-904424310

graingert commented on Misc/NEWS.d/next/Core and Builtins/2021-08-23-19-55-08.bpo-44962.J00ftt.rst:
```diff
@@ -0,0 +1 @@
+Fix a race in WeakSet when two threads attempt to commit the last pending removal. This fixes asyncio.create_task and fixes a data loss in asyncio.run where shutdown_asyncgens is not run
```
```suggestion
Fix a race in WeakKeyDict, WeakValueDict and WeakSet when two threads attempt to commit the last pending removal. This fixes asyncio.create_task and fixes a data loss in asyncio.run where shutdown_asyncgens is not run
```

graingert committed:
Update Misc/NEWS.d/next/Core and Builtins/2021-08-23-19-55-08.bpo-44962.J00ftt.rst

graingert commented on Misc/NEWS.d/next/Core and Builtins/2021-08-23-19-55-08.bpo-44962.J00ftt.rst:
```diff
@@ -0,0 +1 @@
+Fix a race in WeakKeyDict, WeakValueDict and WeakSet when two threads attempt to commit the last pending removal. This fixes asyncio.create_task and fixes a data loss in asyncio.run where shutdown_asyncgens is not run
```
```suggestion
Fix a race in WeakKeyDictionary, WeakValueDictionary and WeakSet when two threads attempt to commit the last pending removal. This fixes asyncio.create_task and fixes a data loss in asyncio.run where shutdown_asyncgens is not run
```

graingert committed:
Update Misc/NEWS.d/next/Core and Builtins/2021-08-23-19-55-08.bpo-44962.J00ftt.rst

ambv commented on Lib/weakref.py:
```diff
@@ -384,13 +391,18 @@ def _commit_removals(self):
+                try:
+                    del d[key]
+                except KeyError:
+                    pass
```
Nesting here is unnecessary since the `except` block returns.
```suggestion

            try:
                del d[key]
            except KeyError:
                pass
```

graingert committed:
Update Lib/weakref.py

Co-authored-by: Łukasz Langa <lukasz@langa.pl>

ambv commented on Lib/weakref.py:
```diff
@@ -120,13 +120,17 @@ def remove(wr, selfref=ref(self), _atomic_removal=_remove_dead_weakref):
+            except IndexError:
+                return
+            else:
+                _remove_dead_weakref(d, key)
```
The reason `self` fields are interned here isn't so much performance but to preserve behavior at interpreter finalization (e.g. if any weakref operation is made as part of `atexit` registrations). This is why there's an alias here:
https://github.com/python/cpython/blob/28db1f61f20352c02e4ae1518e5aeb6505df3045/Lib/weakref.py#L106

I would use the same local alias here too. Other places you touched use built-in `del` or a `discard` that *is* interned.

ambv replied:
More context: https://bugs.python.org/issue29519

ambv commented on Lib/weakref.py:
```diff
@@ -384,11 +391,16 @@ def _commit_removals(self):
             try:
-                del d[l.pop()]
+                del d[key]
             except KeyError:
                 pass
```
Observation: it kind of weirded me out there has to be a `KeyError` guard here. Even in the old code once you get something popped out of the list, you should be able to address the data key with it, no? There should be no duplicates in `self._pending_removals`!

Well, after some digging I found this isn't about duplicates. This is about a race where one `remove()` call hits the object while it's still `_iterating` (registering the key in `_pending_removals`), and concurrently another `remove()` call coming just a bit later when `_iterating` isn't set anymore (performing an eager removal).

In fact, while `_remove_dead_weakref` is called without a `try:except:` fence, it internally ignores raised KeyErrors. Similarly, the WeakSet is using `discard`.

There's some relevant info in https://bugs.python.org/issue21173 and https://bugs.python.org/issue28427.

ambv commented on Lib/weakref.py:
```diff
@@ -120,13 +120,17 @@ def remove(wr, selfref=ref(self), _atomic_removal=_remove_dead_weakref):
 
     def _commit_removals(self):
-        l = self._pending_removals
+        pop = self._pending_removals.pop
```
```suggestion
        pop = self._pending_removals.pop
        _atomic_removal = _remove_dead_weakref
```

ambv commented on Lib/weakref.py:
```diff
@@ -120,13 +120,17 @@ def remove(wr, selfref=ref(self), _atomic_removal=_remove_dead_weakref):
+            except IndexError:
+                return
+            else:
+                _remove_dead_weakref(d, key)
```
```suggestion
                _atomic_removal(d, key)
```

graingert committed:
cache _atomic_removal in WeakValueDictionary._commit_removals

see also https://bugs.python.org/issue29519

graingert committed:
remove redundant else clauses

graingert commented on Lib/weakref.py:
```diff
@@ -398,11 +398,11 @@ def _commit_removals(self):
                 key = pop()
             except IndexError:
                 return
-            else:
```
http://pylint.pycqa.org/en/latest/technical_reference/features.html?highlight=no-else-return%20(R1705)#refactoring-checker-messages

ambv reviewed (approved):
Normally we request tests for bug fixes but in the case of race conditions, they tend to be flaky.
